src/include/OpenImageIO/paramlist.h

```diff
@@ -306,6 +302,12 @@ class OIIO_UTIL_API ParamValue {
     ustring get_ustring(int maxsize = 64) const;
     ustring get_ustring_indexed(int index) const;
 
+    /// Return the total heap memory allocated by the ParamValue.
+    size_t datasize() const noexcept;
+    /// Return the total memory used by the ParamValue, including this
+    /// object itself and any heap memory allocated data.
+    size_t footprint() const noexcept;
+
```

Comment:
Yes, I still feel that decoupling datasize() and footprint is worth it for readbility.
- footprint() will always be `sizeof(Object) + datasize()`, 
- datasize() will only accumulate members datasize() and allocations.

Whenever Object A is part of the footprint of Object B, you only want to account for A.datasize() in the footprint() of Object B to not double count. This is more visible in PR #4323 !

Reply:
I'm not against datasize per se, I was just thinking that if it doesn't serve a distinct purpose, maybe we don't need it?

I think we all know what footprint is -- it's the total memory cost of an object and everything it owns, as near as we can calculate it.

I'm not sure we have a rigorous definition of what datasize really is, though. What does it mean for something like a string (and also ParamValue) that sometimes stores its data within the object itself (so is counted in sizeof(obj)) and sometimes has heap allocations? And what does it mean if the valid data takes up less than the capacity?

There is already a ParamValue::datasize() but it means "how big is the data that this represents", so you could do
```
    memcpy (mybuffer, pv.data(), pv.datasize());
```
It doesn't measure the heap allocation (if it did, it would return 0 for the cases where no heap allocation is needed), nor does it include the unused capacity.

Reply:
Sorry about the mislead, we'd want datasize() to be anything not part of sizeof(T), if that only contains heap allocations we can rename that heapsize() if that makes more sense, but it because it recursively accumulate members datasize() it is sort of indirect allocations.